File: _hyunseo/9184.py
dp = [[[0] * 21 for _ in range(21)] for _ in range(21)]

def w(a, b, c):
    if min(a, b, c) <= 0:
        return 1
    if  max(a,b,c) > 20:
        return w(20, 20, 20)
    if dp[a][b][c] != 0:
        return dp[a][b][c]
    if a < b < c:
        dp[a][b][c] = w(a, b, c - 1) + w(a, b - 1, c - 1) - w(a, b - 1, c)
    else:
        dp[a][b][c] = w(a - 1, b, c) + w(a - 1, b - 1, c) + w(a - 1, b, c - 1) - w(a - 1, b - 1, c - 1)
    return dp[a][b][c]

while True:
    x, y, z = map(int, input().split())
    if x == -1 and y == -1 and z == -1:
        break
    print(f"w({x}, {y}, {z}) = {w(x, y, z)}")

File: _sungyoon/9184.cpp
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pii;
typedef long long ll;
#define endl "\n"
#define MAX 1e9

int dx[] = {0, 1, -1, 0, 1, -1, -1, 1};
int dy[] = {1, 0, 0, -1, -1, 1, -1, 1};
int a, b, c;
int dp[51][51][51];

int solve(int x, int y, int z) {
    if(x <= 0 || y <= 0 || z <= 0) {
        return 1;
    }
    if(x > 20 || y > 20 || z > 20) {
        return solve(20, 20, 20);
    }

    int &ret = dp[x][y][z];
    if(ret != -1) return dp[x][y][z];

    if(x < y && y < z) {
        ret = solve(x, y, z-1) + solve(x, y-1, z-1) - solve(x, y-1, z);
    }
    else {
        ret = solve(x-1, y, z) + solve(x-1, y-1, z) + solve(x-1, y, z-1) - solve(x-1, y-1, z-1);
    }
    return ret;
}

void input() {
    memset(dp, -1, sizeof(dp));
    
    while(true) {
        cin >> a >> b >> c;

        if(a == -1 && b == -1 && c == -1) {
            break;
        }

        cout << "w(" << a << ", " << b << ", " << c << ") " << "= " << solve(a, b, c) << endl;
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

    input();
}
